api/sirsidynix_authentication_provider.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+from gettext import gettext as _
+from typing import TYPE_CHECKING
+from urllib.parse import urljoin
+
+from api.authenticator import BasicAuthenticationProvider, PatronData
+from core.model.configuration import ExternalIntegration
+from core.util.http import HTTP
+
+if TYPE_CHECKING:
+    from requests import Response
+
+
+class SirsiDynixHorizonAuthenticationProvider(BasicAuthenticationProvider):
+    NAME = "SirsiDynix Horizon Authentication"
+    DESCRIPTION = "SirsiDynix Horizon Webservice Authentication"
+    FLOW_TYPE = "http://librarysimplified.org/authtype/sirsidynix-horizon"
+
+    class Keys:
+        APP_ID = "APP_ID"
+        CLIENT_ID = "CLIENT_ID"
+
+    SETTINGS = [
+        {
+            "key": ExternalIntegration.URL,
+            "label": _("Server URL"),
+            "description": _("The external server url."),
+            "required": True,
+            "default": "https://vendor1-sym.sirsidynix.net/ilsws_current/",
+        },
+        {
+            "key": Keys.APP_ID,
+            "label": _("Application ID"),
+            "description": _(
+                "This can be anything, it is used to group all requests to the API."
+            ),
+            "required": True,
+        },
+        {
+            "key": Keys.CLIENT_ID,
+            "label": _("Client ID"),
+            "description": _("The client ID that should be used to identify this CM."),
+            "required": True,
+        },
+        {
+            "key": BasicAuthenticationProvider.TEST_IDENTIFIER,
+            "label": _("Test Identifier"),
+            "description": BasicAuthenticationProvider.TEST_IDENTIFIER_DESCRIPTION_FOR_OPTIONAL_PASSWORD,
+            "required": True,
+        },
+        {
+            "key": BasicAuthenticationProvider.TEST_PASSWORD,
+            "label": _("Test Password"),
+            "description": BasicAuthenticationProvider.TEST_PASSWORD_DESCRIPTION_OPTIONAL,
+        },
+    ]
+
+    def __init__(self, library, integration: ExternalIntegration, analytics=None):
+        super().__init__(library, integration, analytics)
+        self.server_url = integration.url
+        # trailing slash, else urljoin has issues
+        self.server_url = self.server_url + (
+            "/" if not self.server_url.endswith("/") else ""
+        )
+        self.sirsi_client_id = integration.setting(self.Keys.CLIENT_ID).value
+        self.sirsi_app_id = integration.setting(self.Keys.APP_ID).value
+
+    def remote_authenticate(self, username: str, password: str) -> PatronData | bool:
+        """Authenticate this user with the remote server."""
+        data = self.api_patron_login(username, password)
+        if not data:
+            return False
+
+        return PatronData(
```

Comment:
We can probably sub-class `PatronData` so that we can use it to pass the our session information along to the new `_remote_patron_lookup` function without saving it out to the database.